wire-ios-request-strategy/Sources/E2EIdentity/E2eIKeyPackageRotator.swift

```diff
@@ -0,0 +1,162 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireCoreCrypto
+import WireDataModel
+
+// sourcery: AutoMockable
+public protocol E2eIKeyPackageRotating {
+
+    func rotateKeysAndMigrateConversations(
+        e2eIService: E2eIServiceInterface,
+        certificateChain: String
+    ) async throws
+
+}
+
+public class E2eIKeyPackageRotator: E2eIKeyPackageRotating {
+
+    // MARK: - Types
+
+    enum Error: Swift.Error {
+        case noSelfClient
+        case invalidGroupID
+        case invalidIdentity
+    }
+
+    // MARK: - Properties
+
+    private let coreCryptoProvider: CoreCryptoProviderProtocol
+    private let conversationEventProcessor: ConversationEventProcessorProtocol
+    private let context: NSManagedObjectContext
+    private let commitSender: CommitSending
+
+    private var coreCrypto: SafeCoreCryptoProtocol {
+        get throws {
+            try coreCryptoProvider.coreCrypto(requireMLS: true)
+        }
+    }
+
+    // MARK: - Life cycle
+
+    public init(
+        coreCryptoProvider: CoreCryptoProviderProtocol,
+        conversationEventProcessor: ConversationEventProcessorProtocol,
+        context: NSManagedObjectContext,
+        commitSender: CommitSending? = nil
+    ) {
+        self.coreCryptoProvider = coreCryptoProvider
+        self.conversationEventProcessor = conversationEventProcessor
+        self.context = context
+        self.commitSender = commitSender ?? CommitSender(
+            coreCryptoProvider: coreCryptoProvider,
+            notificationContext: context.notificationContext
+        )
+    }
+
+    // MARK: - Interface
+
+    public func rotateKeysAndMigrateConversations(
+        e2eIService: E2eIServiceInterface,
+        certificateChain: String
+    ) async throws {
+
+        // We need to cast this to `WireE2eIdentity` because we only have access
+        // to the protocol it conforms to (WireE2eiIdentityProtocol)
+        guard let identity = e2eIService.e2eIdentity as? WireE2eIdentity else {
+            throw Error.invalidIdentity
+        }
+
+        // Get the rotate bundle from core crypto
+        let rotateBundle = try await coreCrypto.perform {
+            try await $0.e2eiRotateAll(
+                enrollment: identity,
+                certificateChain: certificateChain,
+                newKeyPackageCount: 100
```

Comment:
move this to property

Reply:
do you want to add this as `TODO`?

Reply:
added the property https://github.com/wireapp/wire-ios/pull/793/commits/98fa8015a312a2fe4bb2b34ff638723f2d1a0a15